components/supervisor/pkg/supervisor/tasks.go

```diff
@@ -0,0 +1,412 @@
+// Copyright (c) 2020 TypeFox GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package supervisor
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/gitpod-io/gitpod/common-go/log"
+	csapi "github.com/gitpod-io/gitpod/content-service/api"
+	"github.com/gitpod-io/gitpod/supervisor/api"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/backup"
+	"github.com/gitpod-io/gitpod/supervisor/pkg/terminal"
+)
+
+type runContext struct {
+	contentSource csapi.WorkspaceInitSource
+	headless      bool
+	tasks         []*task
+}
+
+type tasksSubscription struct {
+	updates chan []*api.TaskStatus
+	Close   func() error
+}
+
+func (sub *tasksSubscription) Updates() <-chan []*api.TaskStatus {
+	return sub.updates
+}
+
+func (tm *tasksManager) Subscribe() *tasksSubscription {
+	tm.mu.Lock()
+	defer tm.mu.Unlock()
+
+	if len(tm.subscriptions) > maxSubscriptions {
+		return nil
+	}
+
+	sub := &tasksSubscription{updates: make(chan []*api.TaskStatus, 5)}
+	sub.Close = func() error {
+		tm.mu.Lock()
+		defer tm.mu.Unlock()
+
+		// We can safely close the channel here even though we're not the
+		// producer writing to it, because we're holding mu.
+		close(sub.updates)
+		delete(tm.subscriptions, sub)
+
+		return nil
+	}
+	tm.subscriptions[sub] = struct{}{}
+
+	return sub
+}
+
+type task struct {
+	api.TaskStatus
+	config       TaskConfig
+	command      string
+	prebuildChan *chan bool
+}
+
+type tasksManager struct {
+	config            *Config
+	tasks             map[string]*task
+	subscriptions     map[*tasksSubscription]struct{}
+	mu                sync.RWMutex
+	ready             chan struct{}
+	terminalService   *terminal.MuxTerminalService
+	inWorkspaceHelper *backup.InWorkspaceHelper
+}
+
+func newTasksManager(config *Config, terminalService *terminal.MuxTerminalService, inWorkspaceHelper *backup.InWorkspaceHelper) *tasksManager {
```

Comment:
`InWorkspaceHelper` should be `backup.ContentState`.
`InWorkspaceHelper` just implements this interface in a specific role.